src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs

```diff
@@ -0,0 +1,511 @@
+                     startToken.Parent is AnonymousObjectCreationExpressionSyntax))
+                {
+                    // Since the braces are next to each other the span to format is everything up to the opening brace start.
+                    endPoint = startToken.SpanStart;
+                }
+            }
+
+            var style = documentOptions.GetOption(FormattingOptions.SmartIndent);
+            if (style == FormattingOptions.IndentStyle.Smart)
+            {
+                // Set the formatting start point to be the beginning of the first word to the left 
+                // of the opening brace location.
+                // skip whitespace
+                while (startPoint >= 0 && char.IsWhiteSpace(text[startPoint]))
+                {
+                    startPoint--;
+                }
+
+                // skip token
+                startPoint--;
+                while (startPoint >= 0 && !char.IsWhiteSpace(text[startPoint]))
+                {
+                    startPoint--;
+                }
+            }
+
+            var spanToFormat = TextSpan.FromBounds(Math.Max(startPoint, 0), endPoint);
+            var rules = document.GetFormattingRules(braceFormattingIndentationRules, spanToFormat);
+            var result = Formatter.GetFormattingResult(root, SpecializedCollections.SingletonEnumerable(spanToFormat), document.Project.Solution.Workspace, documentOptions, rules, cancellationToken);
+            if (result == null)
+            {
+                return (ImmutableArray<TextChange>.Empty, closingPoint);
+            }
+
+            var newRoot = result.GetFormattedRoot(cancellationToken);
+            var newClosingPoint = newRoot.GetAnnotatedTokens(s_closingBraceSyntaxAnnotation).Single().SpanStart + 1;
+
+            var textChanges = result.GetTextChanges(cancellationToken).ToImmutableArray();
+            return (textChanges, newClosingPoint);
+
+            static async Task<Document> GetDocumentWithAnnotatedClosingBraceAsync(Document document, int closingBraceEndPoint, CancellationToken cancellationToken)
+            {
+                var originalRoot = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+                var closeBraceToken = originalRoot.FindToken(closingBraceEndPoint - 1);
+                var newCloseBraceToken = closeBraceToken.WithAdditionalAnnotations(s_closingBraceSyntaxAnnotation);
+                var root = originalRoot.ReplaceToken(closeBraceToken, newCloseBraceToken);
+                return document.WithSyntaxRoot(root);
+            }
+        }
+
+        private static ImmutableArray<AbstractFormattingRule> GetBraceIndentationFormattingRules(DocumentOptionSet documentOptions)
+        {
+            var indentStyle = documentOptions.GetOption(FormattingOptions.SmartIndent);
+            return ImmutableArray.Create(BraceCompletionFormattingRule.ForIndentStyle(indentStyle));
```

Comment:
do we adequately test the three different smart indent styles?

Reply:
Yeah we have tests for the cases (see IndentStyle).  If not specified it's the default value (smart)
http://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp.EditorFeatures.UnitTests/AutomaticCompletion/AutomaticBraceCompletionTests.cs,1144